keps/sig-api-machinery/3937-declarative-validation/README.md

```diff
@@ -0,0 +1,1077 @@
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+With this enhancement, Kubernetes Developers will declare API validation using
+IDL tags in Go. For example:
+
+```go
+// staging/src/k8s.io/api/example/v1/types.go
+
+type FizzBuzzSpec struct {
+  // ...
+  //
+  // +maxLength=64
+  Name string `json...`
+  // ...
+}
+
+type FizzBuzz struct {
+  // ...
+  //
+  // +validations=rule:"self.replicas <= self.maxReplicas", messageExpression:"'replicas must be less than maxReplicas value of ' + string(self.maxReplicas)"
+  Spec FizzBuzzSpec `json...`
+}
+```
+
+In this example, both `+maxLength` and `+validations` are IDL tags.
+
+The declarative validation will be used by the kube-apiserver to
+validate API requests.
+
+The declarative validation will also be included in the published OpenAPI:
+
+```yaml
+openAPIV3Schema:
```

Comment:
> Thanks for the doc, excited to see that happen!
> 
> I have a few questions that are mostly all rooting from the same place:
> 
> * Where is this going to happen in the apiserver, before decoding, after decoding?

My plan is to put validation after decoding primarily because we have multiple content types (JSON and proto).

> * Will this use the openapi as the source of truth? Or will the code be generated like we do for defaults?

The prototype uses openapi as source of truth and I plan to implement it that way. But this is an area where I think the implementation can course correct as needed because it won't impact the user visible behavior.

The main complication with using openapi as source of truth is that any validation rules that we want to keep "server side only" will need to be filtered from the openapi before serving it via discovery.

> * Will you decode into unstructured and then do the validation and then decode into built-in type?

Longer term (beta), we may convert straight to the versioned go type and use SMD value wrappers during validation.  The vast majority of write requests never need to be converted away from v1 in request handling.

Short term (1st alpha), we may convert from internal to the versioned type and then validate, probably as unstructured.

> * Or will you run against decoded values? (i.e. null string  becomes empty string for non-nil strings).
>   There may be a few more questions depending on some answers :-)

This is what happens today, so we should be able to migrate all existing validations?